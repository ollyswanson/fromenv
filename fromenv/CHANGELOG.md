# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0](https://github.com/ollyswanson/fromenv/releases/tag/fromenv-v0.1.0) - 2025-08-10

### Added

- Adjust error message based on error count
- Add "path" to error reporting
- Improve diagnostics from unrecognized fields
- requirements

### Fixed

- Update pub struct error message

### Other

- Specify version of fromenv-derive
- Add missing items to manifest
- Use cargo-rdme to interpolate rustdocs
- Add licenses
- Remove `FromEnvError` from public API
- Document usage + motivation
- Include Send + Sync bounds on BoxError
- Rename `ParserResult` to `ParseResult`
- Test error messages + rename ui tests
- Add more trybuild tests
- Scaffold trybuild tests
- Global rename to `FromEnv`
