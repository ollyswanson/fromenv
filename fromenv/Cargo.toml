[package]
name = "fromenv"
version = "0.1.0"
edition = "2024"

[dependencies]
fromenv-derive = { path = "../fromenv-derive" }

[dev-dependencies]
insta = "1.43.1"
temp-env = "0.3.6"
secrecy = "0.10.3"
