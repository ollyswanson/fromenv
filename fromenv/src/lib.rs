mod error;
mod parser;

#[doc(hidden)]
#[path = "private.rs"]
pub mod __private;

pub use error::{FromEnvError, FromEnvErrors};
pub use fromenv_derive::FromEnv;
pub use parser::ParserResult;

type BoxError = Box<dyn std::error::Error + 'static>;
