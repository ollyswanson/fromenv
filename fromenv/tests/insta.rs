//! These set of tests are supposed to provide an example for combining the
//! `requirements` method with insta to document the environment variables
//! needed by an application.

#[test]
fn snapshot_as_documentation() {
    use foo::bar::AppConfig;
    // simulate module structure.
    #[allow(unused)]
    mod foo {
        pub mod bar {
            use std::net::SocketAddr;

            use fromenv::FromEnv;

            use super::baz::KafkaConfig;

            #[derive(FromEnv)]
            pub struct AppConfig {
                #[env(from, default = "postgres://postgres@postgres/postgres", with = into)]
                pub database_url: String,
                #[env(from, default = "127.0.0.1:3000")]
                pub socket_addr: SocketAddr,
                #[env(nested)]
                pub kafka: KafkaConfig,
            }
        }

        pub mod baz {
            use fromenv::FromEnv;

            #[derive(FromEnv)]
            pub struct KafkaConfig {
                #[env(from = "KAFKA_BROKER")]
                pub broker: String,
            }
        }
    }

    insta::assert_snapshot!(AppConfig::requirements())
}
