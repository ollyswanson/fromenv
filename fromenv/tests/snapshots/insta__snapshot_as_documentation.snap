---
source: fromenv/tests/insta.rs
expression: "AppConfig::requirements()"
---
DATABASE_URL=postgres://postgres@postgres/postgres
SOCKET_ADDR=127.0.0.1:3000
KAFKA_BROKER=
